contrib/windows-cmake/CMakeLists.txt

```diff
@@ -3,48 +3,105 @@ project(hwloc
+# Configure dynamically based on platform capabilities
+include(CheckIncludeFile)
+include(CheckSymbolExists)
+include(CheckCSourceCompiles)
+
+check_include_file("dirent.h" HAVE_DIRENT_H)
+check_include_file("unistd.h" HAVE_UNISTD_H)
+check_include_file("malloc.h" HAVE_MALLOC_H)
+check_include_file("memory.h" HAVE_MEMORY_H)
+
+check_symbol_exists(mkstemp "stdlib.h" HAVE_MKSTEMP)
+check_symbol_exists(memalign "malloc.h" HAVE_MEMALIGN)
+
+check_symbol_exists(strncasecmp "strings.h" HAVE_STRNCASECMP)
+if(MSVC AND HAVE_STRNCASECMP)
+    set(hwloc_strncasecmp 1)
+    set(hwloc_strncasecmp_fcn strncasecmp)
+else()
+    set(hwloc_strncasecmp 0)
+    set(hwloc_strncasecmp_fcn strncmp)
+endif()
+
 
+set(SIZEOF_VOID_P ${CMAKE_SIZEOF_VOID_P})
 # disable x86 entirely by default
-string(REPLACE "#define HWLOC_X86_32_ARCH 1" "/* #undef HWLOC_X86_32_ARCH */" PRIVATE_CONFIG_H "${PRIVATE_CONFIG_H}")
-string(REPLACE "#define HWLOC_X86_64_ARCH 1" "/* #undef HWLOC_X86_64_ARCH */" PRIVATE_CONFIG_H "${PRIVATE_CONFIG_H}")
-# and now reenable x86-36 or x86-64 if detected
-if (CMAKE_SYSTEM_PROCESSOR STREQUAL "AMD64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+set(HWLOC_X86_32_ARCH)
+set(HWLOC_X86_64_ARCH)
+set(HWLOC_HAVE_X86_CPUID 1)
+if (CMAKE_SYSTEM_PROCESSOR MATCHES "(^AMD64$|^x86_64$)")
     # "AMD64" on Windows, "x86_64" on Linux
-    string(REPLACE "/* #undef HWLOC_X86_64_ARCH */" "#define HWLOC_X86_64_ARCH 1" PRIVATE_CONFIG_H "${PRIVATE_CONFIG_H}")
-elseif (CMAKE_SYSTEM_PROCESSOR STREQUAL "x86" OR CMAKE_SYSTEM_PROCESSOR MATCHES "i.86")
+    set(HWLOC_X86_64_ARCH 1)
+elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "(^x86$|i.86)")
     # "x86" on Windows, "i.86" on Linux
-    string(REPLACE "/* #undef HWLOC_X86_32_ARCH */" "#define HWLOC_X86_32_ARCH 1" PRIVATE_CONFIG_H "${PRIVATE_CONFIG_H}")
+    set(HWLOC_X86_32_ARCH 1)
+else()
+    set(HWLOC_HAVE_X86_CPUID 0)
 endif()
 
+check_c_source_compiles("int main(void) {int cpuinfo[4]; __cpuidex(cpuinfo,0,0); return 0;}"
+HWLOC_HAVE_MSVC_CPUIDEX
+)
+
 # the following lines are disabled until we are sure they are safe with old build environmentx
 # - snprintf() returned broken values in the past, hwloc detects it during configure (see 7a4ee26510c06b55fc04aaccbfa18d0ca3b87198)
-#   string(REPLACE "#define HAVE_DECL_SNPRINTF 0" "#define HAVE_DECL_SNPRINTF 1" PRIVATE_CONFIG_H "${PRIVATE_CONFIG_H}")
+#   set(HAVE_DECL_SNPRINTF 1)
 # - strtoull() had some issues in the past (see 9559bd08b79ef63dce45df87fb7f875b73ecb512)
-#   string(REPLACE "#define HAVE_DECL_STRTOULL 0" "#define HAVE_DECL_STRTOULL 1" PRIVATE_CONFIG_H "${PRIVATE_CONFIG_H}")
+#   set(HAVE_DECL_STRTOULL 1)
+
+# --- optional external libraries
+find_package(LibXml2)
+set(HWLOC_HAVE_LIBXML2)
+if(LibXml2_FOUND)
+    set(HWLOC_HAVE_LIBXML2 1)
+endif()
+
+find_package(OpenCL)
+set(HWLOC_HAVE_OPENCL)
+if(OpenCL_FOUND)
+    set(HWLOC_HAVE_OPENCL 1)
+endif()
```

Comment:
Optional dependencies are a bit of annoying because the behavior might optionally vary and depend even on build order in a package manager. Better have a option variable called `HWLOC_WITH_LIBXML2` and do a `find_package(... REQUIRED)` call. Same for OpenCL and Cuda

Reply:
0ff77aaa7 does this now